internal/webhooks/cluster.go

```diff
@@ -551,11 +551,36 @@ func ValidateClusterForClusterClass(cluster *clusterv1.Cluster, clusterClass *cl
 	return allErrs
 }
 
+// validateClusterClassExistsAndIsReconciled will try to get the ClusterClass referenced in the Cluster. If it does not exist or is not reconciled it will add a warning.
+// In any other case it will return an error.
+func (webhook *Cluster) validateClusterClassExistsAndIsReconciled(ctx context.Context, newCluster *clusterv1.Cluster) (*clusterv1.ClusterClass, admission.Warnings, error) {
```

Comment:
I assume these non-nil error outcomes are just a normal part of the async reconciliation flow (maybe you could call them known, unavoidable race conditions?)

Or is a existent cluster w/ a 404 ClusterClass or in-progress ClusterClass truly a rare event?

In any event, just curious why/how often these events happen, and if they are regular, temporary side-effects during initial ClusterClass reconciliation, why we would classify these as warnings (and not simply identify these intermittent outcomes and then enqueue).

Reply:
We can't really requeue (at least not on our side). 


These races seem likely to impact. For example in the following, probably common, scenarios:

- if a ClusterClass and Cluster are part of the same yaml
- if a ClusterClass is using external variables defined in a runtime extension
- if there's any error in the ClusterClass e.g. external templates not existing

These aren't errors so we don't introduce a strict order of operations - i.e. a dependency on a ClusterClass existing before a Cluster can be created. All of this is also being validated inside the Clusterclass reconcile, so the warnings are being added to improve the user experience in these cases and actual user error - e.g. typos in the class name.

Reply:
Just to avoid. This PR modifies the Cluster webhook. So this code is not run during reconciliation and we can only return warnings or errors from webhooks. There is no info 😀